### ISSUE TITLE
k_mem_pool ignores 75% of pool memory when n_max = 1

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1693]
**_Reported by Andy Ross:_**

In the process of investigating GH-1612 (a portability problem, not a bug) it was discovered that the mem_pool allocation has a bug.

The recursive _build_block_set assembly macro is supposed to arrange things such that each "level" of the pool's "block sets" manages blocks that are 4x smaller than the previous level (and thus there should be 4x more of these blocks).  But due to the way the "round up" logic works, when the specified number of max blocks is exactly one, the second level will have blocks that are 1/4 the size, but it will still only have *one* of those blocks, ignoring the rest of the pool memory.  And further levels are computed based on this size, and have the same misfeature.

Basically: you can only use the last 75% of the pool if you allocate one big block.

Workaround is just to use an n_max of 2 or more, but that then prevents the use of the pool for a single large allocation.

(Imported from Jira ZEP-1845)

**Comment by nashif** [comment]
**_by Andy Ross:_**

Fix up at: https://gerrit.zephyrproject.org/r/11985

**Comment by nashif** [comment]
**_by Andy Ross:_**

Heh, uh.. never mind.  It all works, it's just a little crazy internally.  What I thought I saw as an incorrect quad_block count on the second and later levels turns out to have been the correct values, as the quad block isn't itself a unit of allocation but instead an implicit array of four.  Which makes those ones fine, and only the *first* one "invalid" (because it's too small, being only a single block and not four).  But that's fine because there's a special case in the implementation to handle that.

Not a bug.  Closing.  I'll still submit the same patch (fixed to get the counts correct) for the portability bug in GH-1612 though.
